Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -16,130 +16,313 @@ import SymbolKit
+private let question    = UTF8.CodeUnit(ascii: "?")
+private let colon       = UTF8.CodeUnit(ascii: ":")
+
+private extension ContiguousArray<UTF8.CodeUnit>.SubSequence {
+     /// Checks if the UTF-8 string looks like a tuple with comma separated values.
     ///
     /// This is used to remove redundant parenthesis around expressions.
     func isTuple() -> Bool {
-        guard first == "(", last == ")", contains(",")else { return false }
+        guard first == openParen, last == closeParen else { return false }
         var depth = 0
         for char in self {
             switch char {
-            case "(": 
+            case openParen:
                 depth += 1
-            case ")": 
+            case closeParen:
                 depth -= 1
-            case "," where depth == 1:
+            case comma where depth == 1:
                 return true
-            default: 
+            default:
                 continue
             }
         }
         return false
     }
-    
-    /// Transforms the string to apply Swift syntactic sugar.
+}
+
+private extension ContiguousArray<UTF8.CodeUnit> {
+    /// Transforms the UTF-8 string to apply Swift syntactic sugar.
     ///
-    /// The transformed string has all occurrences of `Array<Element>`, `Optional<Wrapped>`, and `Dictionary<Key,Value>` replaced with `[Element]`, `Wrapped?`, and `[Key:Value]`.
-    func withSwiftSyntacticSugar() -> String {
-        // If this type uses known Objective-C types, return the original type name.
-        if contains("NSArray<") || contains("NSDictionary<") {
-            return String(self)
-        }
-        // Don't need to do any processing unless this type contains some string that type name that Swift has syntactic sugar for.
-        guard contains("Array<") || contains("Optional<") || contains("Dictionary<") else {
-            return String(self)
+    /// - Parameter markers: Locations of `A<Element>`, `O<Wrapped>`, and `D<Key,Value>` (truncated above) to replace with `[Element]`, `Wrapped?`, and `[Key:Value]`.
+    mutating func applySwiftSyntacticSugar(markers: ContiguousArray<Int>) {
+        assert(!markers.isEmpty, "This is a private helper function and it's the callers responsibility to check if it needs to be called or not.")
+        
+        // Iterating over the UTF-8 string once to find all the balancing angle brackets (`<` and `>`)
+        var markedAngleBracketPairs = ContiguousArray<(open: Int, close: Int)>()
+        markedAngleBracketPairs.reserveCapacity(32) // Some temporary space to work with.
+        
+        var angleBracketStack = ContiguousArray<Int>()
+        angleBracketStack.reserveCapacity(32) // Some temporary space to work with.
+        
+        for index in indices {
+            switch self[index] {
+            case openAngle:
+                angleBracketStack.append(index)
+            case closeAngle:
+                let open = angleBracketStack.removeLast()
```

Comment:
I tested this in the debugger, and saw a fatal error here:

```
Fatal error: Can't remove last element from an empty collection
```

This occurred with declarations that included a return value - for example when this was passed into `applySwiftSyntacticSugar `:

```
(D<Key, Value>.Element)  -> Bool
```

It looks like this code assumes the angle brackets will always be balanced, which is not the case.

Reply:
Wow. I can't believe either of the projects I tested this with had any closure parameters 🤦‍♂️.

I fixed this in https://github.com/swiftlang/swift-docc/pull/1047/commits/8e3be0f0d0ec542546d77181d23b211c40a99765 and added a handful of increasingly complicated closure type tests.

I also changed the code to return the unmodified string in release builds instead of crashing if there are other cases of unbalanced brackets that this doesn't account for. (This asserts in debug builds so that we can find these unexpected cases and investigate them.)